content/blog/typescript-type-generics/index.md

```diff
@@ -0,0 +1,340 @@
+
+```typescript
+// Because this is an `async` function, we want to wrap the returned type value in a Promise
+async function logTheValue<ItemT>(item: ItemT): Promise<{loggedValue: string, original: ItemT, err: Error | undefined}> {
+	// ... Function body here
+}
+```
+
+Alternatively, we could utilize another feature of generics — the ability to pass the type value of the generic manually — and make an interface with a generic and do so there:
+
+```typescript
+interface LogTheValueReturnType<originalT> {
+  loggedValue: string;
+  original: originalT;
+  err: Error | undefined;
+}
+
+// Notice how we're even wrapping that interface in a built-in type with a generic argument for `Promise`!
+async function logTheValue<ItemT>(item: ItemT): Promise<LogTheValueReturnType<ItemT>> {
+	// ... Function body here
+}
+```
+
+WIth these few features, we're able to utilize much of the functionality of generics. 
+
+However, I know I haven't answered what the `<>` really is for. Well, much like type variables, there's also the ability to pass types as "type arguments" when generics are applied to a function.
+
+An example of this would be a syntax like this:
+
+```typescript
+logTheValue<number>(3);
+```
+
+# Non-Function Generics {#non-function-generics}
+
+As you saw before with the `LogTheValueReturnType` interface — functions aren't the only ones with generics. In addition to using them within functions and interfaces, you can also use them in classes. 
+
+Classes with generics can be particularly helpful for data structures like this:
+
+```typescript
+// DataType might want to be a base64 encoded string, a buffer, or an IntArray
+class ImageType<DataType> {
+	data: DataType;
+  height: number;
+  width: number;
+
+  constructor(data: DataType, height: number, width: number) {
+		this.data = data;
+		this.height = height;
+		this.width = width
+	};
+}
+
+function handleImageBuffer(img: ImageType<Buffer>) {}
+```
+
+Type generics in classes can be used as method argument and property types alike.
+
+There's also the ability to use generics within `type` definitions:
+
+```typescript
+interface ImageType<DataType> {
+	data: DataType;
+  height: number;
+  width: number;	
+}
+
+interface ImageConvertMethods<DataType> {
+  // This is the typing of a method. It will take a prop of the generic type and return the generic type
+  toPNG: (data: DataType) => DataType;
+  toJPG: (data: DataType) => DataType;
+}
+
+
+type ImageTypeWithConvertMethods<DataType> = ImageType<DataType> & ImageConvertMethods<DataType>
+```
+
+# Okay, but why-_er_? {#polymorphic-functions}
+
+Well, "I want to see your credentials" person you, type generics enable us to do things like provide typings for **polymorphic functions**. _Polymorphic functions are functions that can accept a myriad of different types and handle them differently._
```

Comment:
I understand you're jokingly referring to the reader, but the particular phrasing `"I want to see your credentials" person you` is confusing because there are three subjects that I can't quite piece together.

Reply:
Would `Well, you "I want to see your credentials" person you` help?

I wanted to avoid doing `Mr.` or `Ms.`/etc as that opens a frustrating can of assumed gender worms. I might just remove this problem all together and rephrase it

Reply:
How's this?

```suggestion
My my, you don't seem to take my word for it when I tell you that type generics are useful. That's alright, I suppose; After all, doubt while learning can lead to some great questions! 😉

Type generics enable us to do things like provide typings for **polymorphic functions**. _Polymorphic functions are functions that can accept a myriad of different types and handle them differently._
```